doc/_extensions/table_from_rows.py

```diff
@@ -126,29 +110,114 @@ def run(self):
-            source = self.state_machine.input_lines.source(
-                 self.lineno - self.state_machine.input_offset - 1)
-            source_dir = os.path.dirname(os.path.abspath(source))
-            sample_yaml_rows = self._rows_from_sample_yaml(source_dir)
+class TableFromSampleYaml(TableFromRows):
+
+    option_spec = {'header': directives.unchanged}
+
+    @staticmethod
+    def _merge_boards(boards):
+        """Converts pairs of similar board types to one table entry."""
+
+        merge_pairs = [
+            ("nrf9160dk_nrf9160", "nrf9160dk_nrf9160_ns", "nrf9160dk_nrf9160_and_ns"),
+            ("nrf5340dk_nrf5340_cpuapp", "nrf5340dk_nrf5340_cpuapp_ns", "nrf5340dk_nrf5340_cpuapp_and_cpuapp_ns"),
+            ("nrf5340dk_nrf5340_cpuapp", "nrf5340dk_nrf5340_cpunet", "nrf5340dk_nrf5340_cpuapp_and_cpunet"),
+            ("thingy53_nrf5340_cpuapp", "thingy53_nrf5340_cpuapp_ns", "thingy53_nrf5340_cpuapp_and_cpuapp_ns"),
+        ]
+
+        for b1, b2, merged in merge_pairs:
+            if b1 in boards and b2 in boards:
+                boards.remove(b1)
+                boards.remove(b2)
+                boards.add(merged)
+
+    def _rows_from_sample_yaml(self, path):
+        """Search for a sample.yaml file and return a union of all relevant boards.
+
+        Boards are retrieved from the integration_platforms sections in the file.
+        If the file is not found and the document folder is named "doc", the
+        parent folder is searched. Should a sample.yaml still not be found, all
+        subfolders are searched and every found sample.yaml is used.
+
+        nrf51 and qemu boards are ignored."""
+
+        rel_path = os.path.relpath(path, self.env.srcdir).lstrip('..' + os.path.sep)
+        dir_path = Path(self.config.table_from_rows_base_dir) / rel_path
+
+        sample_yamls = []
+        if (dir_path / 'sample.yaml').exists():
+            sample_yamls.append(dir_path / 'sample.yaml')
+        elif rel_path.endswith('doc'):
+            parent_path = dir_path.parent / 'sample.yaml'
+            if parent_path.exists():
+                sample_yamls.append(parent_path)
         else:
-            sample_yaml_rows = []
+            sample_yamls += list(dir_path.glob('**/sample.yaml'))
 
-        if len(rows) == 0 and len(sample_yaml_rows) == 0:
-            raise self.severe('No "rows" content or "sample-yaml-rows" '
-                              'provided for "%s" ' % self.name)
+        if not sample_yamls:
+            raise self.severe(f'"{path}" not found')
+
+        boards = set()
+        for sample_yaml_path in sample_yamls:
+            try:
+                with open(sample_yaml_path) as sample_yaml:
+                    data = yaml.safe_load(sample_yaml)
+            except FileNotFoundError:
+                raise self.severe(f'"{sample_yaml_path}" not found')
+            except yaml.YAMLError:
+                raise self.severe(f'"{sample_yaml_path}" is an invalid YAML file')
```

Comment:
note: these may be obviated, they won't be catched and produce an error anyway. YAMLError may potentially include syntax error information in the traceback (hidden here).